snap/gadget.go

```diff
@@ -234,6 +256,221 @@ func ReadGadgetInfo(info *Info, classic bool) (*GadgetInfo, error) {
+			return fmt.Errorf("invalid structure %s: %v", fmtIndexAndName(idx, s.Name), err)
+		}
+		if s.Name != "" {
+			structureNames[s.Name] = true
+		}
+	}
+
+	// validate relative offsets that reference other structures by name
+	for idx, s := range vol.Structure {
+		if s.OffsetWrite.RelativeTo != "" && !structureNames[s.OffsetWrite.RelativeTo] {
+			return fmt.Errorf("structure %v refers to an unknown structure %q",
+				fmtIndexAndName(idx, s.Name), s.OffsetWrite.RelativeTo)
+		}
+
+		if !s.IsBare() {
+			// content relative offset only possible if it's a bare structure
+			continue
+		}
+		for cidx, c := range s.Content {
+			if c.OffsetWrite.RelativeTo != "" && !structureNames[c.OffsetWrite.RelativeTo] {
+				return fmt.Errorf("structure %v, content %v refers to an unknown structure %q",
+					fmtIndexAndName(idx, s.Name), fmtIndexAndName(cidx, c.Image), c.OffsetWrite.RelativeTo)
+			}
+		}
+	}
+	return nil
+}
+
+func validateStructureType(s string, vol *GadgetVolume) error {
+	// Type can be one of:
+	// - "mbr" (backwards compatible)
+	// - "bare"
+	// - [0-9A-Z]{2} - MBR type
+	// - hybrid ID
+	//
+	// Hybrid ID is 2 hex digits of MBR type, followed by 36 GUUID
+	// example: EF,C12A7328-F81F-11D2-BA4B-00A0C93EC93B
+
+	schema := vol.Schema
+	if schema == "" {
+		schema = "gpt"
+	}
+
+	if s == "" {
+		return errors.New(`type is not specified`)
+	}
+
+	if s == "bare" {
+		// unknonwn blob
+		return nil
+	}
+
+	if s == "mbr" {
+		// backward compatibility for type: mbr
+		return nil
+	}
+
+	var isGPT, isMBR bool
+
+	idx := strings.IndexRune(s, ',')
+	if idx == -1 {
+		// just ID
+		switch {
+		case validTypeID.MatchString(s):
+			isMBR = true
+		case validGUUID.MatchString(s):
+			isGPT = true
+		default:
+			return fmt.Errorf("invalid format of type ID")
+		}
+	} else {
+		// hybrid ID
+		code := s[:idx]
+		guid := s[idx+1:]
+		if len(code) != 2 || len(guid) != 36 || !validTypeID.MatchString(code) || !validGUUID.MatchString(guid) {
+			return fmt.Errorf("invalid format of hybrid type ID")
+		}
+	}
+
+	if schema != "gpt" && isGPT {
+		// type: <uuid> is only valid for GPT volumes
+		return fmt.Errorf("GUID structure ID %q with non-GPT schema %q", s, vol.Schema)
+	}
+	if schema != "mbr" && isMBR {
+		return fmt.Errorf("MBR structure ID %q with non-MBR schema %q", s, vol.Schema)
+	}
+
+	return nil
+}
+
+func validateRole(vs *VolumeStructure, vol *GadgetVolume) error {
+	if vs.Type == "bare" {
+		if vs.Role != "" && vs.Role != "mbr" {
+			return fmt.Errorf("confclicting role/type: %q/%q", vs.Role, vs.Type)
+		}
+	}
+	vsRole := vs.Role
+	if vs.Type == "mbr" {
+		// backward compatibility
+		vsRole = "mbr"
+	}
+
+	switch vsRole {
+	case "system-data":
+		if vs.Label != "" && vs.Label != "writable" {
+			return fmt.Errorf(`role %q must have an implicit label or "writable", not %q`, vs.Role, vs.Label)
+		}
+	case "mbr":
+		if vs.Size > 446 {
+			return errors.New("mbr structures cannot be larger than 446 bytes")
+		}
+		if vs.Offset != 0 {
```

Comment:
do I understand correctly that OffsetWrite can be set in this case?

Reply:
I don't see anything in ubuntu-image that would prevent that. OTOH, that wouldn't be particularly useful either.